csharp/ql/src/utils/model-generator/internal/CaptureTypeBasedSummaryModels.qll

```diff
@@ -0,0 +1,231 @@
+ */
+private predicate source(DotNet::Callable callable, string input, TypeParameter tp) {
+  exists(DelegateType dt, int position, Type return, string access |
+    delegate(callable, dt, position) and
+    return = dt.getReturnType() and
+    access = getAccess(callable, return, tp) and
+    input = "Argument[" + position + "].ReturnValue" + access
+  )
+}
+
+/**
+ * Holds if `input` is a models as data string representation of, how a
+ * value of type `tp` (or a generic parameterized over `tp`)
+ * can be provided as input to `callable`.
+ * This includes
+ * (1) The implicit synthetic field(s) of the declaring type of `callable`.
+ * (2) The parameters of `callable`.
+ * (3) Any delegate parameters of `callable`.
+ */
+private predicate input(DotNet::Callable callable, string input, TypeParameter tp) {
+  input = implicit(callable, tp)
+  or
+  parameter(callable, input, tp)
+  or
+  source(callable, input, tp)
+}
+
+/**
+ * Holds if `callable` returns a value of type `tp` (or a generic parameterized over `tp`) and `output`
+ * is a models as data string representation of, how data is routed to the return.
+ */
+private predicate returns(DotNet::Callable callable, TypeParameter tp, string output) {
+  exists(Type return, string access | return = callable.getReturnType() |
+    access = getAccess(callable, return, tp) and
+    output = "ReturnValue" + access
+  )
+}
+
+/**
+ * Holds if `callable` has a delegate parameter that accepts a value of type `tp`
+ * and `output` is the models as data string representation of, how data is routed to
+ * the delegate parameter.
+ */
+private predicate sink(DotNet::Callable callable, TypeParameter tp, string output) {
+  exists(DelegateType dt, int position, Type t, Parameter p |
+    delegate(callable, dt, position) and
+    p = dt.getAParameter() and
+    t = p.getType() and
+    t = tp and
+    output = "Argument[" + position + "]" + ".Parameter[" + p.getPosition() + "]"
+  )
+}
+
+/**
+ * Holds if `output` is a models as data string representation of, how values of type `tp`
+ * (or generics parameterized over `tp`) can be routed.
+ * This includes
+ * (1) The implicit synthetic field(s) of the declaring type of `callable`.
+ * (2) The return of `callable`.
+ * (3) Any delegate parameters of `callable`.
+ */
+private predicate output(DotNet::Callable callable, TypeParameter tp, string output) {
+  output = implicit(callable, tp)
+  or
+  returns(callable, tp, output)
+  or
+  sink(callable, tp, output)
+}
+
+/**
+ * A class of callables that are relevant generating summaries for based
+ * on the Theorems for Free approach.
+ */
+class TypeBasedFlowTargetApi extends Specific::TargetApiSpecific {
+  TypeBasedFlowTargetApi() { Specific::isRelevantForTypeBasedFlowModels(this) }
+
+  /**
+   * Gets the string representation of all type based summaries inspired by
+   * the Theorems for Free approach.
+   *
+   * Basic example signatures could be
```

Comment:
I don't quite understand this; I think it would be better with concrete C# examples.

Reply:
Yes, that is a good idea!